Sources/Public/BacktraceCredentials.swift

```diff
@@ -28,3 +28,92 @@ import Foundation
+        case .endpoint(let endpoint, _):
+            return try parseUniverseName(endpoint.absoluteString)
+        }
+    }
+
+    private func parseUniverseName(_ urlString: String) throws -> String {
+        let backtraceSubmitUrl = "https://submit.backtrace.io/"
+
+        if urlString.starts(with: backtraceSubmitUrl) {
+            let universeIndexStart =
+                urlString.index(urlString.startIndex, offsetBy: backtraceSubmitUrl.count)
+            let substring = urlString[universeIndexStart...urlString.index(before: urlString.endIndex)]
+
+            guard var universeIndexEnd = substring.firstIndex(of: "/") else {
+                throw BacktraceUrlParsingError.invalidInput(urlString)
+            }
+            universeIndexEnd = substring.index(before: universeIndexEnd)
+
+            return String(urlString[universeIndexStart...universeIndexEnd])
+        } else {
+            let backtraceDomain = "backtrace.io"
+            if !urlString.contains(backtraceDomain) {
+                throw BacktraceUrlParsingError.invalidInput(urlString)
+            }
+
+            let url = URL(string: urlString)
+            guard let host = url?.host else {
+                throw BacktraceUrlParsingError.invalidInput(urlString)
+            }
+
+            guard var universeIndexEnd = host.firstIndex(of: ".") else {
+                throw BacktraceUrlParsingError.invalidInput(urlString)
+            }
+            universeIndexEnd = host.index(before: universeIndexEnd)
+
+            return String(host[host.startIndex...universeIndexEnd])
+        }
+    }
+
+    // Using algorithm from backtrace-unity
+    // swiftlint:disable line_length
+    // https://github.com/backtrace-labs/backtrace-unity/blob/553aab2b39c318ff96ebed4bc739bf2c87304649/Runtime/Model/BacktraceConfiguration.cs#L320
+    func getSubmissionToken() throws -> String {
+        switch configuration {
+        case .submissionUrl(let url):
+            let tokenLength = 64
+            let tokenQueryParam = "token="
+            let urlString = url.absoluteString
+
+            if urlString.contains("submit.backtrace.io") {
+                guard var tokenEndIndex = urlString.lastIndex(of: "/") else {
+                    throw BacktraceUrlParsingError.invalidInput(urlString)
+                }
+                tokenEndIndex = urlString.index(before: tokenEndIndex)
+
+                let tokenStartIndex = urlString.index(after: urlString.index(tokenEndIndex, offsetBy: -(tokenLength - 1)))
+
+                return String(urlString[tokenStartIndex...tokenEndIndex])
+            } else {
+                guard let tokenQueryParamRange = urlString.range(of: tokenQueryParam) else {
+                    throw BacktraceUrlParsingError.invalidInput(urlString)
+                }
+
+                let tokenStartIndex = tokenQueryParamRange.upperBound
+                print(urlString[tokenStartIndex...urlString.index(before: urlString.endIndex)])
```

Comment:
Prints should never be used in the production codebase. Please remove it or use a proper logger if this log is useful for debugging.

Reply:
Will remove this, thank you!